src/connection.rs

```diff
@@ -91,370 +14,160 @@ pub fn parse_mavlink_string(buf: &[u8]) -> String {
-                    }
-                };
-
-                self.buf_cache.extend(&self.buf[0..len]);
-
-                let mut start = 0;
-                loop {
-                    match self.buf[start..len].iter().position(|&x| x == 0xfe) {
-                        Some(i) => {
-                            if start + i + 8 > self.buf.len() {
-                                break;
-                            }
-
-                            let len = self.buf[start + i + 1] as usize;
-
-                            if start + i + 8 + len > self.buf.len() {
-                                break;
-                            }
-
-                            let packet;
-                            {
-                                let pktbuf = &self.buf[(start + i)..(start + i + 8 + len)];
-                                packet = MavPacket::new(pktbuf);
-
-                                // println!("ok {:?}", pktbuf);
-
-                                if !packet.check_crc() {
-                                    // println!("failed CRC!");
-                                    start += i + 1;
-                                    continue;
-                                }
-                            }
-
-                            // handle packet
-                            if let Some(pkt) = packet.parse() {
-                                self.dispatch(pkt);
-                            }
-
-                            // change this
-                            start += i + 8 + len;
-                        }
-                        None => {
-                            break;
-                        }
-                    }
+impl MavConnection for Udp {        
+    fn recv(&self) -> io::Result<MavMessage> {
+        let mut l = self.read.lock().unwrap();
+        let lock = &mut *l;
+        loop {
+            if lock.recv_buf.position() as usize >= lock.recv_buf.get_ref().len() {
+                lock.recv_buf.set_position(0);
+                let mut buf = lock.recv_buf.get_mut();
+                buf.resize(64*1024, 0);
+                trace!("Waiting for UDP packet");
+                let (len, src) = try!(lock.socket.recv_from(&mut buf));
+                buf.truncate(len);
+                trace!("UDP received {} bytes", len);
+                
+                if self.server {
+                    self.write.lock().unwrap().dest = Some(src);
                 }
-
-                self.buf_cache = self.buf_cache.split_off(start);
-
-                // Re-register the socket with the event loop. The current
-                // state is used to determine whether we are currently reading
-                // or writing.
-                // self.reregister(event_loop);
             }
-            _ => panic!("Received unknown token"),
+            
+            if let Ok((_, m)) = read(&mut lock.recv_buf) {
+                return Ok(m);
+            } else {
+                // Drop the remainder of the packet
```

Comment:
Note: we probably want to comment that this won't work if we're processing UDP packets packed with several MAVLink messages as solo has outputted (externally) in the past. I can't see any scenario where it impacts our driver code today though.